graylog2-server/src/main/java/org/graylog/plugins/views/migrations/V20191125144500_MigrateDashboardsToViewsSupport/V20191125144500_MigrateDashboardsToViews.java

```diff
@@ -0,0 +1,225 @@
+        );
+
+        final Map<String, Set<String>> newWidgetMapping = new HashMap<>(newViewWidgets.size());
+        final BiConsumer<String, String> recordWidgetMapping = (String viewWidgetId, String searchTypeId) -> newWidgetMapping
+                .merge(viewWidgetId, Collections.singleton(searchTypeId), Sets::union);
+
+        final DateTime createdAt = dashboard.createdAt();
+
+        final Set<SearchType> newSearchTypes = newViewWidgets.stream()
+                .flatMap(viewWidget -> createSearchType(viewWidget, recordWidgetMapping).stream())
+                .collect(Collectors.toSet());
+        final Query newQuery = Query.create(randomUUIDProvider.get(), RelativeRange.create(300),"", newSearchTypes);
+        final Set<Query> newQueries = Collections.singleton(newQuery);
+        final Search newSearch = Search.create(randomObjectIdProvider.get(), newQueries, dashboard.creatorUserId(), createdAt);
+
+        final ViewState newViewState = ViewState.create(
+                Titles.ofWidgetTitles(newWidgetTitles).withQueryTitle(dashboard.title()),
+                newViewWidgets,
+                newWidgetMapping,
+                newViewWidgetPositions
+        );
+
+        final View newView = View.create(
+                dashboard.id(),
+                View.Type.DASHBOARD,
+                dashboard.title(),
+                "This dashboard was migrated automatically.",
+                dashboard.description(),
+                newSearch.id(),
+                Collections.singletonMap(newQuery.id(), newViewState),
+                Optional.ofNullable(dashboard.creatorUserId()),
+                createdAt
+        );
+
+        recordMigratedDashboardIds.accept(dashboard.id());
+        recordMigratedWidgetMap.accept(migratedWidgetIds);
+
+        return new AbstractMap.SimpleEntry<>(newView, newSearch);
+    }
+
+    private Set<SearchType> createSearchType(ViewWidget viewWidget, BiConsumer<String, String> recordWidgetMapping) {
+        final Set<SearchType> searchTypes = viewWidget.toSearchTypes(randomUUIDProvider);
+        searchTypes.forEach(searchType -> recordWidgetMapping.accept(viewWidget.id(), searchType.id()));
+        return searchTypes;
+    }
+
+    private Set<ViewWidget> migrateWidget(Widget widget,
+                                     BiConsumer<String, String> recordMigratedWidgetIds,
+                                     BiConsumer<String, String> recordWidgetTitle) {
+        final Set<ViewWidget> viewWidgets = widget.toViewWidgets(this.randomUUIDProvider);
+
+        viewWidgets.forEach(viewWidget -> {
+            recordWidgetTitle.accept(viewWidget.id(), widget.description());
+            recordMigratedWidgetIds.accept(widget.id(), viewWidget.id());
+        });
+
+        return viewWidgets;
+    }
+
+    private Map<String, ViewWidgetPosition> migrateWidgetPositions(Dashboard dashboard, Map<String, Set<String>> migratedWidgetIds, Set<ViewWidget> viewWidgets) {
```

Comment:
I would suggest to extract this entire method to `Dashboard::extractWidgetPositions`. Actually I would extract all mapping logic out if this class and keep only the higher level-stuff related to collecting the migrated ids and tracking `MigrationCompleted`. This would make it easier to grok that everything else is "just" mapping code.

It would also facilitate writing separate unit tests for the mapping logic, if desired.

Reply:
Talked to @dennisoelkers abouth this. I'm ok with leaving it to save time. After all we'll hopefully not be looking at these classes too often in the future.